This adds `parseutils.parseSize`, an inverse to `strutils.formatSize`

c-blake opened the pull request:
which has existed since 2017.

It is useful for parsing the compiler's own output logs (like SuccessX) or many other scenarios where "human readable" units have been chosen. The doc comment and tests explain accepted syntax in detail.

Big units lead to small numbers, often with a fractional part, but we parse into an `int64` since that is what `formatSize` stringifies and this is an inverse over partial function slots.  Although metric prefixes z & y for zettabyte & yottabyte are accepted, these will saturate the result at `int64.high` unless the qualified number is a small fraction.  This should not be much of a problem until such sizes are common (at which point another overload with the parse result either `float64` or `int128` could be added).

Tests avoids `test()` because of a weakly related static: test() failure as mentioned in https://github.com/nim-lang/Nim/pull/21325. This is a more elemental VM failure.  As such, it needs its own failure exhibition issue that is a smaller test case.  (I am working on that, but unless there is a burning need to `parseSize` at compile-time before run-time it need not hold up this PR.)

This PR replaces https://github.com/nim-lang/Nim/pull/21325 which got a little git-mangled.

c-blake committed:
This adds `parseutils.parseSize`, an inverse to `strutils.formatSize`
which has existed since 2017.

It is useful for parsing the compiler's own output logs (like SuccessX)
or many other scenarios where "human readable" units have been chosen.
The doc comment and tests explain accepted syntax in detail.

Big units lead to small numbers, often with a fractional part, but we
parse into an `int64` since that is what `formatSize` stringifies and
this is an inverse over partial function slots.  Although metric
prefixes z & y for zettabyte & yottabyte are accepted, these will
saturate the result at `int64.high` unless the qualified number is a
small fraction.  This should not be much of a problem until such sizes
are common (at which point another overload with the parse result
either `float64` or `int128` could be added).

Tests avoids `test()` because of a weakly related static: test() failure
as mentioned in https://github.com/nim-lang/Nim/pull/21325. This is a
more elemental VM failure.  As such, it needs its own failure exhibition
issue that is a smaller test case.  (I am working on that, but unless
there is a burning need to `parseSize` at compile-time before run-time
it need not hold up this PR.)

c-blake committed:
This worked with `int` but fails with `int64`.  Try for green tests.

juancarlospaco commented on tests/stdlib/tparseutils.nim:
```diff
@@ -55,5 +55,42 @@ proc test() =
+
+block:  # With this included, static: test() crashes the compiler (from a
+        # VM problem with parseSize calling parseFloat).
+  template ass(x) = doAssert(x) # Avoid hint[LineTooLong]
```
But instead of using the `ass` try not using `;` and doing 1 assert per line,
I think the `LineTooLong` hint is disabled by default anyways.

c-blake committed:
Lift 2-result matching into a `checkParseSize` template and format as a
table of input & 2 expected outputs which seems nicer and to address
https://github.com/nim-lang/Nim/pull/21349#pullrequestreview-1294407679

c-blake committed:
Fix (probably) the i386 trouble by using `int64` consistently.

c-blake committed:
Improve documentation by mentioning saturation.

juancarlospaco commented on lib/pure/parseutils.nim:
```diff
@@ -592,6 +592,61 @@ proc parseFloat*(s: openArray[char], number: var float): int {.
+  ##
+  ## **See also:**
+  ## * https://en.wikipedia.org/wiki/Binary_prefix
+  ## * `formatSize module<strutils.html>`_ for formatting
```
Add a `runnableExample` for the Docs?.

c-blake replied:
Agreed.  Thanks!  I have a commit, but I want to wait for these hours long tests to finish before pushing it (and so cancelling what is running).

c-blake replied:
Well, that little `int` -> `int64` thing made i386 CI run fine, as expected.  So, I committed & pushed.  Unless I have a bug in my `runnableExamples`, I think this is looking pretty good.

c-blake committed:
Improve documentation with `runnableExamples` and a little more detail in
the main doc comment based on excellent code review by @juancarlospaco:
https://github.com/nim-lang/Nim/pull/21349#pullrequestreview-1294564155

juancarlospaco commented on lib/pure/parseutils.nim:
```diff
@@ -592,6 +592,71 @@ proc parseFloat*(s: openArray[char], number: var float): int {.
+  ## An optional trailing 'B|b' is ignored but processed.  I.e., you must still
+  ## know if units are bytes | bits or infer this fact via the case of s[^1] (if
+  ## users can even be relied upon to use 'B' for byte and 'b' for bit or have
+  ## that be s[^1]).  Trailing text stops the parse, as usual.
```
```suggestion
  ## that be s[^1]).  Trailing text stops the parse.
```
Too informal.

c-blake replied:
I was of two minds about the whole Trailing text sentence as it is kind of redundant upon the first paragraph.  Maybe drop the whole ~~thing~~ sentence?

c-blake replied:
For what it's worth, in a draft I had said something like "since milli-bytes are not a thing" instead of "fractional information" bit, but I decided that was too informal.  Ultimately, that's all part of explaining why the corruption of metric notation (G -> g, M -> m, K|k, etc.) is so common. 

Anyway, while technically accurate, I'm also unsure "fractional information" is likely to be understood, but I'm not sure of a pithy, non-informal way to phrase it, if you have any ideas.

juancarlospaco commented on lib/pure/parseutils.nim:
```diff
@@ -592,6 +592,71 @@ proc parseFloat*(s: openArray[char], number: var float): int {.
+  ##
+  ## If `alwaysBin==true` then scales are always binary-metric, but e.g. "KiB"
+  ## is still accepted for clarity.  If the value would exceed the range of
+  ## `int64`, `size` is set to `int64.high`.  Supported metric prefix chars
```
```suggestion
  ## `int64`, then `size` is saturated. Supported metric prefix chars
```

c-blake replied:
Well, I suppose the same module does have `parseSaturatedNatural`, but I was thinking being more specific might be helpful.  Maybe first occurrence "saturated to `int64.high`" instead of "set to" and second occurrence only "z & y saturate"?

juancarlospaco commented on lib/pure/parseutils.nim:
```diff
@@ -592,6 +592,71 @@ proc parseFloat*(s: openArray[char], number: var float): int {.
+  ## is still accepted for clarity.  If the value would exceed the range of
+  ## `int64`, `size` is set to `int64.high`.  Supported metric prefix chars
+  ## include k, m, g, t, p, e, z, y (but z & y saturate to int64.high unless
+  ## number is a small fraction).  It is unlikely you will need beyond 'peta'.
```
```suggestion
  ## number is a small fraction).  If values beyond 'peta' are needed then try a Nimble library.
```

Or link a bigint lib in docs.

c-blake replied:
This may be another case where just dropping the questionable sentence entirely is best.  I'm fine with that.

juancarlospaco commented on lib/pure/parseutils.nim:
```diff
@@ -592,6 +592,71 @@ proc parseFloat*(s: openArray[char], number: var float): int {.
+func toLowerAscii(c: char): char =
+  if c in {'A'..'Z'}: char(uint8(c) xor 0b0010_0000'u8) else: c
+
+func parseSize*(s: openArray[char], size: var int64, alwaysBin=false): int =
```
Considering is totally new, should it use `size: out int64` ?. 🤔
Needs a `{.since:   .}` pragma thing.

c-blake replied:
I am kind of hoping to backport it to nim-1.6..

c-blake replied:
Also, all of the other `parseFoo` in the module still just use a `var`... So, there may be a "change them all at the same time" kind of  consistency principle in play.

c-blake committed:
Address some more @juancarlospaco code review concerns.

juancarlospaco commented on lib/pure/parseutils.nim:
```diff
@@ -592,6 +592,71 @@ proc parseFloat*(s: openArray[char], number: var float): int {.
+    doAssert parseSize("64 mib", res) == 6
+    doAssert res == 67108864    # 64 shl 20
+    doAssert parseSize("1G/h", res, true) == 2 # '/' stops parse
+    doAssert res == 1073741824  # 1 shl 30, forced binary metric
```
`runnableExamples` use `assert` in stdlib as convention.
(I use `doAssert` myself but stdlib has that convention).

c-blake replied:
I got that from the other `runnableExamples` in the same module.  So, I was just trying to be stylistically consistent.  I'd say this is another "change all module procs at the same time" kind of question, like `var` vs. `out`.

c-blake committed:
Remove a stray space.

c-blake committed:
Mention milli-bytes in docs to maybe help clarify why wild conventions
are so prone to going case-insensitive-metric.

c-blake commented:
AFAICT, the failing rebuild packages test for chronos here has nothing to do with this PR.

c-blake committed:
Add some parens.
